tests/common/dhcp_relay_utils.py

```diff
@@ -14,3 +22,152 @@ def _is_dhcp_relay_ready():
+    downlink_vlan_iface = dhcp_relay['downlink_vlan_iface']['name']
+    # it can be portchannel or interface, it depends on the topology
+    uplink_portchannels_or_interfaces = dhcp_relay['uplink_interfaces']
+    client_iface = dhcp_relay['client_iface']['name']
+    portchannels = dhcp_relay['portchannels']
+
+    '''
+    If the uplink_portchannels_or_interfaces are portchannels,
+        uplink_interfaces will contains the members of the portchannels
+    If the uplink_portchannels_or_interfaces are not portchannels,
+        uplink_interfaces will equal to uplink_portchannels_or_interfaces
+    '''
+    uplink_interfaces = []
+    common_msg = "comparing {} counters and {} counters, hostname:{}, "
+
+    for portchannel_name in uplink_portchannels_or_interfaces:
+        if portchannel_name in portchannels.keys():
+            uplink_interfaces.extend(portchannels[portchannel_name]['members'])
+            portchannel_counters = query_and_sum_dhcpcom_relay_counters(duthost,
+                                                                        downlink_vlan_iface,
+                                                                        [portchannel_name])
+            members_counters = query_and_sum_dhcpcom_relay_counters(duthost,
+                                                                    downlink_vlan_iface,
+                                                                    portchannels[portchannel_name]['members'])
+            # Compare the portchannel counters with the sum of its members' counters
+            logger.info(common_msg.format(
+                portchannel_name, portchannels[portchannel_name]['members'], duthost.hostname))
+
+            # If the portchannel counters and its members' counters are not equal, yield a warning message
+            compare_result = compare_dhcpcom_relay_counter_values(
+                portchannel_counters, members_counters, error_in_percentage)
+            while msg := next(compare_result, False):
+                logger.warning(
+                    ("Warning found when " + common_msg + ": {}")
+                    .format(portchannel_name, portchannels[portchannel_name]['members'], duthost.hostname, str(msg)))
+
+        else:
+            uplink_interfaces.append(portchannel_name)
+
+    vlan_interface_counter = query_and_sum_dhcpcom_relay_counters(duthost, downlink_vlan_iface, [])
+    client_interface_counter = query_and_sum_dhcpcom_relay_counters(duthost, downlink_vlan_iface, [client_iface])
+    uplink_portchannels_interfaces_counter = query_and_sum_dhcpcom_relay_counters(
+        duthost, downlink_vlan_iface, uplink_portchannels_or_interfaces
+    )
+    uplink_interface_counter = query_and_sum_dhcpcom_relay_counters(duthost, downlink_vlan_iface, uplink_interfaces)
+
+    logger.info(common_msg.format(downlink_vlan_iface, client_iface, duthost.hostname))
+    compare_result = compare_dhcpcom_relay_counter_values(vlan_interface_counter,
```

Comment:
Can we extract line 140 - line 148 into a function and reuse it? Looks like below invoking are duplicated